<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="William Casazza">

<title>mQTL Training Module</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="main-module_files/libs/clipboard/clipboard.min.js"></script>
<script src="main-module_files/libs/quarto-html/quarto.js"></script>
<script src="main-module_files/libs/quarto-html/popper.min.js"></script>
<script src="main-module_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="main-module_files/libs/quarto-html/anchor.min.js"></script>
<link href="main-module_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="main-module_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="main-module_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="main-module_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="main-module_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#objectives" id="toc-objectives" class="nav-link active" data-scroll-target="#objectives">Objectives</a>
  <ul class="collapse">
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link" data-scroll-target="#required-packages">Required packages</a></li>
  </ul></li>
  <li><a href="#quantitative-trait-loci-qtl-basics" id="toc-quantitative-trait-loci-qtl-basics" class="nav-link" data-scroll-target="#quantitative-trait-loci-qtl-basics">Quantitative trait loci (QTL) basics</a>
  <ul class="collapse">
  <li><a href="#match-samples" id="toc-match-samples" class="nav-link" data-scroll-target="#match-samples">Match Samples</a></li>
  <li><a href="#compute-pcs" id="toc-compute-pcs" class="nav-link" data-scroll-target="#compute-pcs">Compute PCs</a></li>
  <li><a href="#write-data" id="toc-write-data" class="nav-link" data-scroll-target="#write-data">Write data</a></li>
  <li><a href="#pc-checking" id="toc-pc-checking" class="nav-link" data-scroll-target="#pc-checking">PC checking</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">mQTL Training Module</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>William Casazza </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="objectives" class="level1">
<h1>Objectives</h1>
<ol type="1">
<li>Basics of molecular quantitative trait locus (molQTL) analysis</li>
<li>Running cis-methylation QTL analysis with the <code>matrixEQTL</code> package</li>
<li>Post-hoc analyses involving molQTL and GWAS summary statistics</li>
</ol>
<section id="required-packages" class="level2">
<h2 class="anchored" data-anchor-id="required-packages">Required packages</h2>
<pre class="{r}"><code>library(tidyverse)
library(data.table)
library(minfi)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
knitr::opts_chunk$set(echo = TRUE)
methy_annot &lt;- as.data.table(getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19))</code></pre>
</section>
</section>
<section id="quantitative-trait-loci-qtl-basics" class="level1">
<h1>Quantitative trait loci (QTL) basics</h1>
<p>Quantitative trait loci are “mapped” by computing the association between SNP genotype and a quantitative trait like DNA methylation (DNAm), gene expression, or some other epigenetic mark.</p>
<pre class="{r}"><code>richs_ancestry &lt;- fread("../RICHS_genotyping/phg001102.v1.RICHS.genotype-calls-matrixfmt.MEGA_Consortium.c1.GRU/raw_data.geno.maf.mind.sex_check.het_filter.ibd_filter.eigenvec")
subject_pheno &lt;- fread("/scratch/st-dennisjk-1/wcasazza/RICHS_genotyping/RICHS_Subject_phenotypes.GRU.txt")
sample_multi &lt;- fread("/scratch/st-dennisjk-1/wcasazza/RICHS_genotyping/RICHS_Sample.MULTI.txt")
fam &lt;- read.delim("/scratch/st-dennisjk-1/wcasazza/RICHS_genotyping/phg001102.v1.RICHS.genotype-calls-matrixfmt.MEGA_Consortium.c1.GRU/plink_for_dbgap.fam",header=FALSE,sep="")
mapping &lt;- fread("RICHS_dbgap_GEO_subject_sample_mapping.csv")
combined_meta &lt;- subject_pheno %&gt;%
    left_join(
        sample_multi
    )%&gt;% filter(SAMPLE_USE == "Array_SNP")
richs_meta &lt;- combined_meta
richs_meta$IID &lt;- richs_meta$SAMPLE_ID
richs_meta_methy &lt;- fread("../RICHS_DNAm/metadata.csv") %&gt;% inner_join(mapping,by=c("sample"= "GSM_SampleID"))
richs_meta &lt;- richs_meta %&gt;% right_join(richs_ancestry,by=c("IID"="V2")) %&gt;% right_join(richs_meta_methy,by="SAMPLE_ID") %&gt;% filter(SAMPLE_ID %in% richs_ancestry$V2)
betas_sub &lt;- fread("processed_DNAm_RICHS_functional_norm.txt")
genotype &lt;- fread("../RICHS_genotyping/phg001102.v1.RICHS.genotype-calls-matrixfmt.MEGA_Consortium.c1.GRU/raw_data.imputed.r2_30.maf_mismatch.traw")</code></pre>
<section id="match-samples" class="level2">
<h2 class="anchored" data-anchor-id="match-samples">Match Samples</h2>
<pre class="{r}"><code>colnames(genotype) &lt;- gsub(".*_","",colnames(genotype))
richs_meta &lt;- richs_meta[IID %in% colnames(genotype) &amp; sample %in% colnames(betas_sub)]
gen_vars &lt;- c(colnames(genotype)[1:6],richs_meta$IID)
to_delete &lt;- colnames(genotype)[!colnames(genotype)%in% gen_vars]
genotype[,c(to_delete):=NULL]
setcolorder(genotype,gen_vars)
methy_vars &lt;- c("cpg",richs_meta$sample)
betas_sub &lt;- betas_sub[,methy_vars,with=FALSE]
</code></pre>
<pre class="{r}"><code>all(colnames(betas_sub)[-c(1)] == richs_meta$sample)
all(colnames(genotype)[-c(1:6)] == richs_meta$IID)</code></pre>
</section>
<section id="compute-pcs" class="level2">
<h2 class="anchored" data-anchor-id="compute-pcs">Compute PCs</h2>
<pre class="{r}"><code>covars &lt;- richs_meta %&gt;% as.data.frame() %&gt;%
                dplyr::select(
                    age,
                    V3,
                    V4,
                    V5,
                    V6,
                    V7,
                    sex
                ) %&gt;%
                mutate(Sex= as.numeric(sex =="M"))%&gt;%
                dplyr::select(-sex) %&gt;%
                as.data.frame()
residual &lt;- residuals(lm(as.matrix(t(betas_sub[!cpg %in% methy_annot[chr %in% c("chrX","chrY")]$Name ,-c(1)]))~., data=covars))
methy_PC &lt;- prcomp(residual, center=T,rank.= 20)</code></pre>
</section>
<section id="write-data" class="level2">
<h2 class="anchored" data-anchor-id="write-data">Write data</h2>
<pre class="{r}"><code>#SNP POS
fwrite(genotype[,.(SNP,CHR=paste0("chr",CHR),POS)],"matrix_eqtl_data/snp_pos.txt",sep="\t", quote=F)
#SNP POS
fwrite(genotype[CHR==21,.(SNP,CHR=paste0("chr",CHR),POS)],"matrix_eqtl_data/snp_pos_chr21.txt",sep = "\t", quote=F)
#SNPs
geno_vars &lt;- colnames(genotype)[c(1,3:6)]
genotype[,c(geno_vars):=NULL]
fwrite(genotype,"matrix_eqtl_data/all_imputed_matrixeQTL.txt",sep = "\t",quote = F)
#Methy POS
fwrite(methy_annot[Name %in% betas_sub$cpg,.(geneid=Name,chr,s1=pos,s2=pos)], "matrix_eqtl_data/probe_pos.txt",sep = "\t",quote=F)
#methy
fwrite(betas_sub,"matrix_eqtl_data/methylation_matrixeQTL.txt",sep="\t",quote=F)


#SNPs
snp_pos &lt;- fread("matrix_eqtl_data/snp_pos.txt")
fwrite(genotype[SNP %in% snp_pos[CHR=="chr21"]$SNP],"matrix_eqtl_data/all_imputed_matrixeQTL_chr21.txt", sep = "\t", quote = F)
#Methy POS
fwrite(methy_annot[Name %in% betas_sub$cpg &amp; chr=="chr21",.(geneid=Name, chr,s1=pos,s2=pos)], "matrix_eqtl_data/probe_pos_chr1.txt", sep = "\t", quote=F)
chr1_cpg &lt;- methy_annot[Name %in% betas_sub$cpg &amp; chr == "chr21",]$Name
#methy
fwrite(betas_sub[cpg %in% chr1_cpg],"matrix_eqtl_data/methylation_matrixeQTL_chr21.txt",sep="\t", quote=F)</code></pre>
<pre class="{r}"><code>fmt &lt;- "matrix_eqtl_data/mQTL_covar_%d_methy_PC.txt"
for(i in 0:20){
    if (i==1){
        tmp &lt;- t(cbind(PC1=methy_PC$x[,1:i],richs_meta %&gt;% as.data.frame() %&gt;%
                dplyr::select(
                    age,
                    V3,
                    V4,
                    V5,
                    V6,
                    V7,
                    sex
                )) %&gt;%
                mutate(Sex= as.numeric(sex =="M"))%&gt;%
                dplyr::select(-sex)) %&gt;%
                as.data.frame()

    }else if (i==0){
          tmp &lt;- t(cbind(richs_meta %&gt;% as.data.frame() %&gt;%
               dplyr::select(
                    age,
                    V3,
                    V4,
                    V5,
                    V6,
                    V7,
                    sex
                )) %&gt;%
                mutate(Sex= as.numeric(sex =="M"))%&gt;%
                dplyr::select(-sex)) %&gt;%
                as.data.frame()
    }else{
        tmp &lt;- t(cbind(methy_PC$x[,1:i],
                richs_meta %&gt;% as.data.frame() %&gt;%
                dplyr::select(
                    age,
                    V3,
                    V4,
                    V5,
                    V6,
                    V7,
                    sex
                )) %&gt;%
                mutate(Sex= as.numeric(sex =="M"))%&gt;%
                dplyr::select(-sex)) %&gt;%

                as.data.frame()

    }
  colnames(tmp) &lt;- richs_meta$IID
    write_delim(
      tmp %&gt;% rownames_to_column(var="id"),
      path=sprintf(fmt,i))
}</code></pre>
</section>
<section id="pc-checking" class="level2">
<h2 class="anchored" data-anchor-id="pc-checking">PC checking</h2>
<pre class="{r}"><code>fmt &lt;- "matrix_eqtl_data/cis_all_impute_mQTL_results_%d_methy_PC_chr21.txt"
chr21_res &lt;- lapply(
  0:20,
  function(i){
    tmp &lt;- fread(sprintf(fmt,i))
    nrow(tmp[`p-value` &lt; (0.05/ .N)])
  }
)</code></pre>
<pre class="{r}"><code>to_plot &lt;- data.frame(PC=0:20,hits=unlist(chr21_res))
ggplot(to_plot,aes(PC,hits)) + geom_point() + geom_smooth() + labs(x="PCs Included",y="mQTL Detected")+ theme_minimal()</code></pre>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>